File: src/compiler/typescript_to_rust/generators/ic_object/functions/reject.ts
import { Rust } from '../../../../../types';

export function generateIcObjectFunctionReject(): Rust {
    return /* rust */ `
        fn _azle_ic_reject(
            _this: &boa_engine::JsValue,
            _aargs: &[boa_engine::JsValue],
            _context: &mut boa_engine::Context,
        ) -> boa_engine::JsResult<boa_engine::JsValue> {
            Ok(
                ic_cdk::api::call::reject(_aargs.get(0).unwrap().as_string().unwrap())
                    .azle_into_js_value(_context)
            )
        }
    `;
}

File: src/compiler/typescript_to_rust/azle_generate/src/body/ic_object/functions/reject.rs
pub fn generate() -> proc_macro2::TokenStream {
    quote::quote! {
        fn reject(
            _this: &boa_engine::JsValue,
            aargs: &[boa_engine::JsValue],
            context: &mut boa_engine::Context,
        ) -> boa_engine::JsResult<boa_engine::JsValue> {
            let message: String = aargs
                .get(0)
                .ok_or_else(|| "An argument for 'message' was not provided".to_js_error(None))?
                .clone()
                .try_from_vm_value(&mut *context)?;

            ic_cdk::api::call::reject(&message)
                .try_into_vm_value(&mut *context)
                .map_err(|vmc_err| vmc_err.to_js_error(None))
        }
    }
}
